apisix/plugins/kafka-logger.lua

```diff
@@ -77,28 +76,70 @@ local function log(premature, conf, log_message)
         end
     end
 
-    broker_config["request_timeout"] = conf.timeout
-    broker_config["max_retry"] = conf.max_retry
-
-    --Async producers will queue logs and push them when the buffer exceeds.
-    if conf.async then
-        broker_config["producer_type"] = "async"
-    end
+    broker_config["request_timeout"] = conf.timeout * 1000
 
     local prod, err = producer:new(broker_list,broker_config)
     if err then
-        core.log.error("failed to identify the broker specified", err)
-        return
+        return nil, "failed to identify the broker specified: " .. err
     end
 
     local ok, err = prod:send(conf.kafka_topic, conf.key, log_message)
     if not ok then
-        core.log.error("failed to send data to Kafka topic", err)
+        return nil, "failed to send data to Kafka topic" .. err
     end
 end
 
+
 function _M.log(conf)
-    return timer_at(0, log, conf, core.json.encode(log_util.get_full_log(ngx)))
+    local entry = log_util.get_full_log(ngx)
+
+    if not entry.route_id then
+        core.log.error("failed to obtain the route id for kafka logger")
+        return
+    end
+
+    local log_buffer = buffers[entry.route_id]
+
+    if log_buffer then
+        log_buffer:push(entry)
+        return
+    end
+
+    -- Generate a function to be executed by the batch processor
+    local func = function(entries, batch_max_size)
+        local data, err
+        if batch_max_size == 1 then
+            data, err = core.json.encode(entries[1]) -- encode as single {}
+        else
+            data, err = core.json.encode(entries) -- encode as array [{}]
+        end
+
+        if not data then
+            return false, 'error occurred while encoding the data: ' .. err
+        end
+
+        return send_kafka_data(conf, data)
+    end
+
+    local config = {
+        name = conf.name,
+        retry_delay = conf.retry_delay,
+        batch_max_size = conf.batch_max_size,
+        max_retry_count = conf.max_retry_count,
+        buffer_duration = conf.buffer_duration,
+        inactive_timeout = conf.inactive_timeout,
+    }
+
+    local err
+    log_buffer, err = batch_processor:new(func, config)
+
+    if not log_buffer then
+        core.log.error("error when creating the batch processor: ", err)
+        return
+    end
+
+    buffers[entry.route_id] = log_buffer
```

Comment:
I have one question:

If there are no more new logs to be sent later, how to release the `log_buffer` object?
If the log_buffer is useless, how to delete it from `buffers`? If we don't delete it, it will make the `buffers` bigger. 

we need to control the max count number of `timer`. the default value is 1024 for each worker process.

https://github.com/openresty/lua-nginx-module#lua_max_pending_timers

Reply:
Yes, @membphis  I was also thinking about this problem, 
For the memory problem:
we need to write a timer function which executes in all active loggers and finds stale buffer objects and remove the variable. Maybe it can run every 30 mins?

For the timer problem:
The buffer by default has an inactive timeout. If we don't send any logs for 30 seconds, by default it will send the existing logs and will not create any additional timers. Due to this fact, we won't be creating timers for inactive buffers. Only active buffers will have a timer running.

If we agree on the memory optimization technique, then I can send the fix in another PR.

A practical bottleneck is if a user configures like 1024 routes and actively using all the routes.
But I assume its not a practical scenario.

Reply:
I have created an issue to remove stale objects from memory:
https://github.com/apache/incubator-apisix/issues/1494

Reply:
@membphis any update on this?